File: shared/typescript/contracts.ts
export type DataSourceType = "csv" | "excel" | "polygon" | "manual";
export type RunMode = "local_compare" | "pine_bridge";
export type StrategyLanguage = "pine" | "python";
export type AccessLevel = "read" | "write";
export type PermissionScope = "single_action" | "session";
export type PermissionTarget = "pine_code" | "python_code" | "run_artifacts";
export type DiffClassification =
  | "data_alignment"
  | "warmup_window"
  | "timeframe_aggregation"
  | "pine_only_function_gap"
  | "python_implementation"
  | "numeric_tolerance";
export type RunLifecycle = "draft" | "running" | "live" | "completed" | "failed";

export interface ColumnMapping {
  timestamp: string;
  open: string;
  high: string;
  low: string;
  close: string;
  volume?: string | null;
}

export interface SessionConfig {
  timezone: string;
  market_calendar?: string | null;
  session_start?: string | null;
  session_end?: string | null;
}

export interface DataSourceConfig {
  type: DataSourceType;
  name: string;
  file_path?: string | null;
  provider?: string | null;
  symbol?: string | null;
  timeframe: string;
  timezone: string;
  mapping?: ColumnMapping | null;
  session: SessionConfig;
  extra: Record<string, unknown>;
}

export interface StrategyPermission {
  read_allowed: boolean;
  write_allowed: boolean;
}

export interface StrategyArtifact {
  language: StrategyLanguage;
  source_code: string;
  name: string;
  declared_outputs: string[];
  permissions: StrategyPermission;
  adapter_metadata: Record<string, unknown>;
}

export interface RunConfig {
  mode: RunMode;
  symbol: string;
  timeframe: string;
  date_from?: string | null;
  date_to?: string | null;
  one_open_position: boolean;
  tolerance: number;
  warmup_bars: number;
  selected_outputs: string[];
  timezone: string;
  companion_dataset_ids: Record<string, string>;
}

export interface CandlePoint {
  timestamp: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume?: number | null;
}

export interface IndicatorPoint {
  timestamp: string;
  value?: number | null;
}

export interface IndicatorSeries {
  name: string;
  pane: string;
  style: Record<string, unknown>;
  warmup_bars: number;
  values: IndicatorPoint[];
}

export interface TradeEvent {
  timestamp: string;
  side: "long_entry" | "long_exit" | "short_entry" | "short_exit";
  price: number;
  qty: number;
  reason: string;
  source_engine: StrategyLanguage;
}

export interface MismatchDetail {
  classification: DiffClassification;
  series_name: string;
  timestamp: string;
  expected?: number | null;
  actual?: number | null;
  delta?: number | null;
  message: string;
  suspected_region?: string | null;
}

export interface ComparisonSummary {
  aligned: boolean;
  total_series: number;
  mismatched_series: number;
  total_trade_events: number;
  mismatched_trade_events: number;
}

export interface ComparisonResult {
  summary: ComparisonSummary;
  series_mismatches: MismatchDetail[];
  trade_mismatches: MismatchDetail[];
  first_mismatch?: MismatchDetail | null;
  unsupported_feature_warnings: string[];
  suggested_next_action?: string | null;
  run_mode?: RunMode | null;
  dataset_id?: string | null;
  live: boolean;
  artifact_refs: string[];
}

export interface DatasetArtifact {
  dataset_id: string;
  name: string;
  source: DataSourceConfig;
  mapping: ColumnMapping;
  symbol?: string | null;
  timeframe: string;
  timezone: string;
  row_count: number;
  columns: string[];
  data_path: string;
  created_at: string;
}

export interface DatasetPreview {
  preview: Record<string, unknown>;
  capabilities: Record<string, unknown>;
  validation_issues: string[];
}

export interface BridgeArtifact {
  artifact_id: string;
  name: string;
  symbol: string;
  timeframe: string;
  source_code?: string | null;
  indicator_series: IndicatorSeries[];
  trade_events: TradeEvent[];
  notes?: string | null;
  created_at: string;
}

export interface DependencyStatusItem {
  name: string;
  available: boolean;
  detail: string;
}

export interface DependencyStatus {
  backend: DependencyStatusItem;
  ollama: DependencyStatusItem;
  tradingview_bridge: DependencyStatusItem;
  market_provider: DependencyStatusItem;
}

export interface RunStatus {
  run_id: string;
  lifecycle: RunLifecycle;
  mode: RunMode;
  symbol: string;
  timeframe: string;
  dataset_id?: string | null;
  dataset_name?: string | null;
  companion_dataset_ids: Record<string, string>;
  bridge_artifact_id?: string | null;
  python_artifact: StrategyArtifact;
  pine_artifact?: StrategyArtifact | null;
  candles: CandlePoint[];
  python_series: IndicatorSeries[];
  pine_series: IndicatorSeries[];
  trade_events: TradeEvent[];
  comparison?: ComparisonResult | null;
  warnings: string[];
  live_progress: number;
  live_total: number;
  created_at: string;
  updated_at: string;
}

export interface LiveBarEvent {
  run_id: string;
  lifecycle: RunLifecycle;
  current_index: number;
  total: number;
  latest_candle?: CandlePoint | null;
  comparison?: ComparisonResult | null;
  updated_at: string;
}

export interface PermissionGrant {
  target: PermissionTarget;
  access: AccessLevel;
  scope: PermissionScope;
  approved: boolean;
  expires_at?: string | null;
  audit_note?: string | null;
  is_active?: boolean;
}

export interface ChatMessage {
  role: "system" | "user" | "assistant" | "tool";
  content: string;
  created_at?: string;
}

export type ChatStatus = "ok" | "fallback" | "error";
export type ChatErrorClass =
  | "ollama_unreachable"
  | "model_missing"
  | "timeout"
  | "cleaned_response_empty"
  | "permission_required";

export interface ChatRequest {
  model: string;
  intent: "analysis" | "explain_diff" | "propose_fix" | "apply_fix";
  messages: ChatMessage[];
  include_targets: PermissionTarget[];
  run_id?: string | null;
}

export interface ChatResponse {
  model: string;
  content: string;
  requires_approval: boolean;
  proposed_patch?: string | null;
  status: ChatStatus;
  error_class?: ChatErrorClass | null;
  fallback_used: boolean;
}

export interface OllamaModelInfo {
  name: string;
  model: string;
  size?: number | null;
  modified_at?: string | null;
  chat_capable: boolean;
}

export type IndicatorCategory = "trend" | "momentum" | "volatility" | "volume" | "custom";

export interface IndicatorLibraryEntry {
  indicator_id: string;
  name: string;
  description: string;
  category: IndicatorCategory;
  pine_code: string;
  python_code: string;
  series_names: string[];
  is_builtin: boolean;
  created_at: string;
  updated_at: string;
}

File: shared/python/contracts.py
from __future__ import annotations

from datetime import UTC, datetime
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, Field, computed_field


class DataSourceType(str, Enum):
    CSV = "csv"
    EXCEL = "excel"
    POLYGON = "polygon"
    MANUAL = "manual"


class RunMode(str, Enum):
    LOCAL_COMPARE = "local_compare"
    PINE_BRIDGE = "pine_bridge"


class StrategyLanguage(str, Enum):
    PINE = "pine"
    PYTHON = "python"


class AccessLevel(str, Enum):
    READ = "read"
    WRITE = "write"


class PermissionScope(str, Enum):
    SINGLE_ACTION = "single_action"
    SESSION = "session"


class PermissionTarget(str, Enum):
    PINE_CODE = "pine_code"
    PYTHON_CODE = "python_code"
    RUN_ARTIFACTS = "run_artifacts"


class DiffClassification(str, Enum):
    DATA_ALIGNMENT = "data_alignment"
    WARMUP_WINDOW = "warmup_window"
    TIMEFRAME_AGGREGATION = "timeframe_aggregation"
    PINE_ONLY_FUNCTION_GAP = "pine_only_function_gap"
    PYTHON_IMPLEMENTATION = "python_implementation"
    NUMERIC_TOLERANCE = "numeric_tolerance"


class RunLifecycle(str, Enum):
    DRAFT = "draft"
    RUNNING = "running"
    LIVE = "live"
    COMPLETED = "completed"
    FAILED = "failed"


class ColumnMapping(BaseModel):
    timestamp: str
    open: str
    high: str
    low: str
    close: str
    volume: str | None = None


class SessionConfig(BaseModel):
    timezone: str = "Asia/Calcutta"
    market_calendar: str | None = None
    session_start: str | None = None
    session_end: str | None = None


class DataSourceConfig(BaseModel):
    type: DataSourceType
    name: str
    file_path: str | None = None
    provider: str | None = None
    symbol: str | None = None
    timeframe: str = "1D"
    timezone: str = "Asia/Calcutta"
    mapping: ColumnMapping | None = None
    session: SessionConfig = Field(default_factory=SessionConfig)
    extra: dict[str, Any] = Field(default_factory=dict)


class StrategyPermission(BaseModel):
    read_allowed: bool = True
    write_allowed: bool = False


class StrategyArtifact(BaseModel):
    language: StrategyLanguage
    source_code: str
    name: str
    declared_outputs: list[str] = Field(default_factory=list)
    permissions: StrategyPermission = Field(default_factory=StrategyPermission)
    adapter_metadata: dict[str, Any] = Field(default_factory=dict)


class RunConfig(BaseModel):
    mode: RunMode
    symbol: str
    timeframe: str
    date_from: datetime | None = None
    date_to: datetime | None = None
    one_open_position: bool = True
    tolerance: float = 1e-6
    warmup_bars: int = 100
    selected_outputs: list[str] = Field(default_factory=list)
    timezone: str = "Asia/Calcutta"
    companion_dataset_ids: dict[str, str] = Field(default_factory=dict)


class CandlePoint(BaseModel):
    timestamp: datetime
    open: float
    high: float
    low: float
    close: float
    volume: float | None = None


class IndicatorPoint(BaseModel):
    timestamp: datetime
    value: float | None = None


class IndicatorSeries(BaseModel):
    name: str
    pane: str = "main"
    style: dict[str, Any] = Field(default_factory=dict)
    warmup_bars: int = 0
    values: list[IndicatorPoint] = Field(default_factory=list)


class TradeEvent(BaseModel):
    timestamp: datetime
    side: Literal["long_entry", "long_exit", "short_entry", "short_exit"]
    price: float
    qty: float
    reason: str
    source_engine: StrategyLanguage


class MismatchDetail(BaseModel):
    classification: DiffClassification
    series_name: str
    timestamp: datetime
    expected: float | None
    actual: float | None
    delta: float | None
    message: str
    suspected_region: str | None = None


class ComparisonSummary(BaseModel):
    aligned: bool
    total_series: int
    mismatched_series: int
    total_trade_events: int
    mismatched_trade_events: int


class ComparisonResult(BaseModel):
    summary: ComparisonSummary
    series_mismatches: list[MismatchDetail] = Field(default_factory=list)
    trade_mismatches: list[MismatchDetail] = Field(default_factory=list)
    first_mismatch: MismatchDetail | None = None
    unsupported_feature_warnings: list[str] = Field(default_factory=list)
    suggested_next_action: str | None = None
    run_mode: RunMode | None = None
    dataset_id: str | None = None
    live: bool = False
    artifact_refs: list[str] = Field(default_factory=list)


class DatasetArtifact(BaseModel):
    dataset_id: str
    name: str
    source: DataSourceConfig
    mapping: ColumnMapping
    symbol: str | None = None
    timeframe: str
    timezone: str
    row_count: int
    columns: list[str] = Field(default_factory=list)
    data_path: str
    created_at: datetime = Field(default_factory=lambda: datetime.now(UTC))


class DatasetPreview(BaseModel):
    preview: dict[str, Any]
    capabilities: dict[str, Any]
    validation_issues: list[str] = Field(default_factory=list)


class BridgeArtifact(BaseModel):
    artifact_id: str
    name: str
    symbol: str
    timeframe: str
    source_code: str | None = None
    indicator_series: list[IndicatorSeries] = Field(default_factory=list)
    trade_events: list[TradeEvent] = Field(default_factory=list)
    notes: str | None = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(UTC))


class DependencyStatusItem(BaseModel):
    name: str
    available: bool
    detail: str


class DependencyStatus(BaseModel):
    backend: DependencyStatusItem
    ollama: DependencyStatusItem
    tradingview_bridge: DependencyStatusItem
    market_provider: DependencyStatusItem


class RunStatus(BaseModel):
    run_id: str
    lifecycle: RunLifecycle
    mode: RunMode
    symbol: str
    timeframe: str
    dataset_id: str | None = None
    dataset_name: str | None = None
    companion_dataset_ids: dict[str, str] = Field(default_factory=dict)
    bridge_artifact_id: str | None = None
    python_artifact: StrategyArtifact
    pine_artifact: StrategyArtifact | None = None
    candles: list[CandlePoint] = Field(default_factory=list)
    python_series: list[IndicatorSeries] = Field(default_factory=list)
    pine_series: list[IndicatorSeries] = Field(default_factory=list)
    trade_events: list[TradeEvent] = Field(default_factory=list)
    comparison: ComparisonResult | None = None
    warnings: list[str] = Field(default_factory=list)
    live_progress: int = 0
    live_total: int = 0
    created_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(UTC))


class LiveBarEvent(BaseModel):
    run_id: str
    lifecycle: RunLifecycle
    current_index: int
    total: int
    latest_candle: CandlePoint | None = None
    comparison: ComparisonResult | None = None
    updated_at: datetime = Field(default_factory=lambda: datetime.now(UTC))


class PermissionGrant(BaseModel):
    target: PermissionTarget
    access: AccessLevel
    scope: PermissionScope
    approved: bool
    expires_at: datetime | None = None
    audit_note: str | None = None

    @computed_field
    @property
    def is_active(self) -> bool:
        if not self.approved:
            return False
        if self.expires_at is None:
            return True
        return self.expires_at >= datetime.now(UTC)


class ChatMessage(BaseModel):
    role: Literal["system", "user", "assistant", "tool"]
    content: str
    created_at: datetime = Field(default_factory=lambda: datetime.now(UTC))


class ChatStatus(str, Enum):
    OK = "ok"
    FALLBACK = "fallback"
    ERROR = "error"


class ChatErrorClass(str, Enum):
    OLLAMA_UNREACHABLE = "ollama_unreachable"
    MODEL_MISSING = "model_missing"
    TIMEOUT = "timeout"
    CLEANED_RESPONSE_EMPTY = "cleaned_response_empty"
    PERMISSION_REQUIRED = "permission_required"


class ChatRequest(BaseModel):
    model: str
    intent: Literal["analysis", "explain_diff", "propose_fix", "apply_fix"]
    messages: list[ChatMessage]
    include_targets: list[PermissionTarget] = Field(default_factory=list)
    run_id: str | None = None


class ChatResponse(BaseModel):
    model: str
    content: str
    requires_approval: bool = False
    proposed_patch: str | None = None
    status: ChatStatus = ChatStatus.OK
    error_class: ChatErrorClass | None = None
    fallback_used: bool = False


class OllamaModelInfo(BaseModel):
    name: str
    model: str
    size: int | None = None
    modified_at: datetime | None = None
    chat_capable: bool = True


class IndicatorCategory(str, Enum):
    TREND = "trend"
    MOMENTUM = "momentum"
    VOLATILITY = "volatility"
    VOLUME = "volume"
    CUSTOM = "custom"


class IndicatorLibraryEntry(BaseModel):
    indicator_id: str
    name: str
    description: str = ""
    category: IndicatorCategory = IndicatorCategory.CUSTOM
    pine_code: str
    python_code: str
    series_names: list[str] = Field(default_factory=list)
    is_builtin: bool = False
    created_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
